### ISSUE TITLE
`jacfwd` shape error when used for constant function in conjunction with `vmap`

### ISSUE THREAD

**Issue by mihaelacr** [google/jax#3357]
To reproduce:
```
def constant_fn(x):
  return 9.
fn = vmap(constant_fn)
ones_tensor = np.ones((3, 3), dtype=jnp.float32)
print(jax.jacfwd(fn)(ones_tensor))```
```

Version which works:
```
def constant_fn(x):
  return jnp.array(9.)
  
fn = vmap(constant_fn)
ones_tensor = np.ones((3, 3), dtype=jnp.float32)
print(jax.jacfwd(fn)(ones_tensor))```
```

Note that this works with `jacrev` or when `vmap` is not used (by directly asking the jacobian of a constant function).

**Comment by mattjj** [comment]
It turns out @apaszke recently found and fixed this in #3222 ! But that PR hasn't been merged yet. I'll copy his code into a separate patch to close this.

**Comment by mattjj** [comment]
This was fixed in #3222 !
